File: src/main/java/de/kevcodez/ecommerce/parser/impl/BonPrixParser.java
package de.kevcodez.ecommerce.parser.impl;

import static java.util.Collections.singletonList;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import org.jsoup.nodes.Document;
import org.jsoup.select.Elements;

import de.kevcodez.ecommerce.parser.domain.image.ImageDto;
import de.kevcodez.ecommerce.parser.domain.image.ImageVariant;
import de.kevcodez.ecommerce.parser.domain.price.Discount;
import de.kevcodez.ecommerce.parser.downloader.WebsiteSourceDownloader;

public class BonPrixParser extends JsoupProductParser {

    private static final Pattern PATTERN_IMG_DIMENSIONS = Pattern.compile("\\/(\\d+)x(\\d+)\\/");

    public BonPrixParser(WebsiteSourceDownloader websiteSourceDownloader) {
        super(websiteSourceDownloader);
    }

    @Override
    List<String> supportedDomains() {
        return singletonList("bonprix.de");
    }

    @Override
    String parseExternalId(Document document) {
        return document.select("div#product-page").attr("data-product-ordernumber");
    }

    @Override
    String parseTitle(Document document) {
        return document.select("h1.product-name").text();
    }

    @Override
    String parseDescription(Document document) {
        return document.select("p.product-information-full-description").text();
    }

    @Override
    BigDecimal parseCurrentPrice(Document document) {
        String priceAsString = document.select("div#offer > span.price").attr("content");
        return new BigDecimal(priceAsString);
    }

    @Override
    String parseCurrencyCode(String url, Document document) {
        return "EUR";
    }

    @Override
    Discount parseDiscount(BigDecimal currentPrice, Document document) {
        String formerPrice = document.select("span.price.former-price").text();

        if (formerPrice.isEmpty()) {
            return null;
        }

        String oldPriceAsString = formerPrice.replace("€", "").replace(",", ".");

        return Discount.of(new BigDecimal(oldPriceAsString), currentPrice);
    }

    @Override
    List<ImageDto> parseImages(Document document) {
        Elements imageWrapperElements = document.select("div#carousel_product_look div.image-wrapper");

        List<ImageDto> images = new ArrayList<>();

        imageWrapperElements.forEach(element -> {
            ImageDto image = new ImageDto();

            String dataImageSrc = element.attr("data-image-src");
            String dataZoomImageSrc = element.attr("data-zoom-image-src");
            String dataPreviewImageSrc = element.attr("data-preview-image-src");

            image.addVariant(buildImageVariantFromUrl(dataImageSrc));
            image.addVariant(buildImageVariantFromUrl(dataZoomImageSrc));
            image.addVariant(buildImageVariantFromUrl(dataPreviewImageSrc));

            images.add(image);
        });

        return images;
    }

    private ImageVariant buildImageVariantFromUrl(String imageSource) {
        Matcher matcher = PATTERN_IMG_DIMENSIONS.matcher(imageSource);
        if (matcher.find()) {
            int height = Integer.parseInt(matcher.group(1));
            int width = Integer.parseInt(matcher.group(2));

            return ImageVariant.builder()
                .url("https:" + imageSource)
                .height(height)
                .width(width)
                .build();
        }

        throw new IllegalArgumentException("Error parsing images with url " + imageSource);
    }

}

File: src/main/kotlin/de/kevcodez/ecommerce/parser/impl/BonPrixParser.kt
package de.kevcodez.ecommerce.parser.impl

import de.kevcodez.ecommerce.parser.domain.image.Image
import de.kevcodez.ecommerce.parser.domain.image.ImageVariant
import de.kevcodez.ecommerce.parser.domain.price.Discount
import de.kevcodez.ecommerce.parser.downloader.WebsiteSourceDownloader
import de.kevcodez.ecommerce.parser.exception.ParserException
import org.jsoup.nodes.Document
import java.math.BigDecimal
import java.util.*
import java.util.Collections.singletonList
import java.util.regex.Pattern

class BonPrixParser(websiteSourceDownloader: WebsiteSourceDownloader) : JsoupProductParser(websiteSourceDownloader) {

    override fun supportedDomains(): List<String> {
        return singletonList("bonprix.de")
    }

    override fun parseExternalId(document: Document): String {
        return document.select("div#product-page").attr("data-product-ordernumber")
    }

    override fun parseTitle(document: Document): String {
        return document.select("h1.product-name").text()
    }

    override fun parseDescription(document: Document): String {
        return document.select("p.product-information-full-description").text()
    }

    override fun parseCurrentPrice(document: Document): BigDecimal {
        val priceAsString = document.select("div#offer > span.price").attr("content")
        return BigDecimal(priceAsString)
    }

    override fun parseCurrencyCode(url: String, document: Document): String {
        return "EUR"
    }

    override fun parseDiscount(currentPrice: BigDecimal, document: Document): Optional<Discount> {
        var discount: Discount? = null

        val formerPrice = document.select("span.price.former-price").text()

        if (!formerPrice.isEmpty()) {
            val oldPriceAsString = formerPrice.replace("€", "").replace(",", ".")

            discount = Discount.of(BigDecimal(oldPriceAsString), currentPrice)
        }

        return Optional.ofNullable(discount)
    }

    override fun parseImages(document: Document): List<Image> {
        val imageWrapperElements = document.select("div#carousel_product_look div.image-wrapper")

        val images = ArrayList<Image>()

        imageWrapperElements.forEach({ element ->
            val image = Image()

            val dataImageSrc = element.attr("data-image-src")
            val dataZoomImageSrc = element.attr("data-zoom-image-src")
            val dataPreviewImageSrc = element.attr("data-preview-image-src")

            image.addVariant(buildImageVariantFromUrl(dataImageSrc))
            image.addVariant(buildImageVariantFromUrl(dataZoomImageSrc))
            image.addVariant(buildImageVariantFromUrl(dataPreviewImageSrc))

            images.add(image)
        })

        return images
    }

    private fun buildImageVariantFromUrl(imageSource: String): ImageVariant {
        val matcher = PATTERN_IMG_DIMENSIONS.matcher(imageSource)
        if (matcher.find()) {
            val height = Integer.parseInt(matcher.group(1))
            val width = Integer.parseInt(matcher.group(2))

            return ImageVariant(
                    url = IMAGE_PREFIX + imageSource,
                    height = height,
                    width = width)
        }

        throw ParserException("Error parsing images with url " + imageSource)
    }

    companion object {

        private val PATTERN_IMG_DIMENSIONS = Pattern.compile("""\/(\d+)x(\d+)\/""")

        private val IMAGE_PREFIX = "https:"
    }

}
